### PR TITLE
Markdown JSON Mode

### PR DESCRIPTION
**Pull request by Anmol6**
Addresses Issue #235 

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `Circle` class for enhanced image processing.
  - Added a new mode `MD_JSON` for extended functionality in handling responses.

- **Enhancements**
  - Improved the `draw_circle` function to include image encoding to base64.

- **Bug Fixes**
  - Adjusted the `response` logic for better accuracy and performance.

- **Tests**
  - Modified existing assertions and added new tests to ensure reliability of the new features.

### CODE REVIEW

File: instructor/patch.py [hunk 1]
```diff
@@ -79,13 +79,21 @@ def handle_response_model(
...
+                # if it is not, add a system message to the beginning
+                message = f"""Make sure that your response to any message matches the json_schema below,
+                        do not deviate at all: \n{response_model.model_json_schema()['properties']}
+                        """
+            else:
+                message = f"""
+                    As a genius expert, your task is to understand the content and provide 
+                    the parsed objects in json that match the following json_schema (do not deviate at all and its okay if you cant be exact). Use a
+                    ```json<JSON>``` block to provide your response:\n
+                    {response_model.model_json_schema()['properties']}
```
Comment (jxnl) [inline instructor/patch.py:95]:
```suggestion
                    {response_model.model_json_schema()['properties']}
```json
```

Comment (jxnl) [reply instructor/patch.py:95]: this and we can add a `stop="```"` in kwargs so this supports streaming natively and we dont need the json parsing

File: instructor/function_calls.py [hunk 2]
```diff
@@ -237,6 +239,14 @@ def from_response(
                 context=validation_context,
                 strict=strict,
             )
+        elif mode == Mode.MD_JSON:
+            pattern = r'```(?:json)?(.*?)```'
+            json_content = re.findall(pattern, message.content, re.DOTALL)[0].strip()
+            return cls.model_validate_json(
+                json_content.strip(),
```
Comment (jxnl) [inline instructor/function_calls.py:246]: it should just be content. and the ````(?:json)?(.*?)```` should be part of the decoding by starting the message with ```json

File: instructor/patch.py (hunk 1 above)
Comment (Anmol6) [reply instructor/patch.py:95]: should we append the ```json to the user query?

**Comment by jxnl** [comment]
added some tests and found that reask does not work cause it'll apologize ands hit.

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for UserDetails
E             Invalid JSON: expected value at line 1 column 1 [type=json_invalid, input_value='Apologies for the confus...ven JSON schema is:\n\n', input_type=str]
E               For further information visit https://errors.pydantic.dev/2.5/v/json_invalid
```

**Comment by jxnl** [comment]
maybe in patching add a note that like if you want to try the vision model it only supports md_mode, and add a

```
!!! warning "Experimental"

     This is not recommended, and may not be supported in the future, this is left so support vision models
````

**Review by jxnl — CHANGES_REQUESTED** [review]
looks good, can we update docs and add a warning?

**Review by jxnl — APPROVED** [review]

**Pull request merged**